Mathlib/Algebra/Lie/CartanExists.lean

```diff
@@ -0,0 +1,380 @@
+
+/-- Let `x` and `y` be elements of a Lie `R`-algebra `L`, and `M` a Lie module over `M`.
+Then the characteristic polynomials of the family of endomorphisms `⁅r • y + x, _⁆` of `M`
+have coefficients that are polynomial in `r : R`.
+In other words, we obtain a polynomial over `R[X]`
+that specializes to the characteristic polynomial of `⁅r • y + x, _⁆` under the map `X ↦ r`.
+This polynomial is captured in `lieCharpoly R M x y`. -/
+private noncomputable
+def lieCharpoly : Polynomial R[X] :=
+  letI bL := chooseBasis R L
+  letI bM := chooseBasis R M
+  (LinearMap.polyCharpoly (LieHom.toLinearMap φ) bL).map <| RingHomClass.toRingHom <|
+    MvPolynomial.aeval fun i ↦ C (bL.repr y i) * X + C (bL.repr x i)
+
+lemma lieCharpoly_monic : (lieCharpoly R M x y).Monic :=
+  (polyCharpoly_monic _ _).map _
+
+lemma lieCharpoly_natDegree : (lieCharpoly R M x y).natDegree = finrank R M := by
+  rw [lieCharpoly, (polyCharpoly_monic _ _).natDegree_map, polyCharpoly_natDegree]
+
+variable {R} in
+lemma lieCharpoly_map_eval (r : R) :
+    (lieCharpoly R M x y).map (evalRingHom r) = (φ (r • y + x)).charpoly := by
+  rw [lieCharpoly, map_map]
+  set b := chooseBasis R L
+  have aux : (fun i ↦ (b.repr y) i * r + (b.repr x) i) = b.repr (r • y + x) := by
+    ext i; simp [mul_comm r]
+  simp_rw [← coe_aeval_eq_evalRingHom, ← AlgHom.comp_toRingHom, MvPolynomial.comp_aeval,
+    map_add, map_mul, aeval_C, Algebra.id.map_eq_id, RingHom.id_apply, aeval_X, aux,
+    MvPolynomial.coe_aeval_eq_eval, polyCharpoly_map_eq_charpoly, LieHom.coe_toLinearMap]
+
+lemma lieCharpoly_coeff_natDegree (i j : ℕ) (hij : i + j = finrank R M) :
+    ((lieCharpoly R M x y).coeff i).natDegree ≤ j := by
+  classical
+  rw [← mul_one j, lieCharpoly, coeff_map]
+  apply MvPolynomial.aeval_natDegree_le
+  · apply (polyCharpoly_coeff_isHomogeneous φ (chooseBasis R L) _ _ hij).totalDegree_le
+  intro k
+  apply Polynomial.natDegree_add_le_of_degree_le
+  · apply (Polynomial.natDegree_C_mul_le _ _).trans
+    simp only [natDegree_X, le_rfl]
+  · simp only [natDegree_C, zero_le]
+
+end engel_le_engel
+
+end CommRing
+
+section Field
+
+variable {K L : Type*}
+variable [Field K] [LieRing L] [LieAlgebra K L] [Module.Finite K L]
+
+open FiniteDimensional LieSubalgebra Module.Free Polynomial
+
+open Cardinal LieModule LieSubalgebra Polynomial engel_le_engel in
+/-- Let `L` be a Lie algebra of dimension `n` over a field `K` with at least `n` elements.
+Given a Lie subalgebra `U` of `L`, and an element `x ∈ U` such that `U ≤ engel K x`.
+Suppose that `engel K x` is minimal amongst the Engel subalgebras `engel K y` for `y ∈ U`.
+Then `engel K x ≤ engel K y` for all `y ∈ U`.
+
+Lemma 2 in [barnes1967]. -/
+lemma engel_le_engel (hLK : finrank K L ≤ #K)
+    (U : LieSubalgebra K L) (x : U) (hUx : U ≤ engel K (x : L))
+    (hmin : ∀ y : U, engel K (y : L) ≤ engel K (x : L) → engel K (y : L) = engel K (x : L))
+    (y : U) :
+    engel K (x : L) ≤ engel K (y : L) := by
```

Comment:
How about introducing the following somewhere in the order theory basics:
```lean
def IsMinimumOfMinimal {ι α : Type*} [Preorder α] (f : ι → α) : Prop :=
    ∀ i, (∀ j, f j ≤ f i → f j = f i) → ∀ j, f i ≤ f j
```
and then restating this as:
```suggestion
lemma isMinimumOfMinimal_engel_of_finrank_le (hLK : finrank K L ≤ #K) (U : LieSubalgebra K L) :
    IsMinimumOfMinimal <| engel K ∘ U.incl := by
```

Reply:
I think it does. For example if you start like this then the proof works with no modification:
```
lemma isMinimumOfMinimalOn_engel_of_finrank_le
    (hLK : finrank K L ≤ #K) (U : LieSubalgebra K L) :
    IsMinimumOfMinimalOn {x : U | U ≤ engel K (x : L)} (engel K ∘ U.incl) := by
  intro x (hUx : U ≤ engel K (x : L)) hmin y
  change ∀ y : U, engel K (y : L) ≤ engel K (x : L) → engel K (y : L) = engel K (x : L) at hmin
  change engel K (x : L) ≤ engel K (y : L)
```

Reply:
Weird, I don't know what I was trying this afternoon. But you are absolutely right!

Reply:
I've been playing a bit more with the statement. What do you think of the following version:
```lean
lemma engel_le_engel (hLK : finrank K L ≤ #K) (U : LieSubalgebra K L)
    (E : {engel K x | x ∈ U}) (hUE : U ≤ E) (hE : IsMin E) :
    IsBot E := by
```
To me it seems even closer to the informal statement.

Reply:
Done